### ISSUE TITLE
Object literals of "or" union types sometime allow wrong types.

### ISSUE THREAD

**Issue by Mike96Angelo** [microsoft/TypeScript#29327]
**TypeScript Version:**  3.3.0-dev.20190108

**Search Terms:**
- object literals
- or union 

**Code**

```ts
type A = {
  type: 'a' | 'b'
  kind: 1 | 2
}

type B = {
  type: 'a'
  kind: 1
  bOnly: number
}

let obj: A | B

obj = {
  type: 'b',
  kind: 2,
  bOnly: 90, // rightly complains that 'bOnly' is not a part of type 'A'
}

obj = {
  type: 'a',
  kind: 2, // does not complain that '2' cannot be assigned to 'B.kind'
  bOnly: 90, // does not complain that 'bOnly' is not a part of type 'A'
}
```

**Expected behavior:**
The compiler should state: 
- object `{type: 'a', kind: 2, bOnly: number}` is not assignable to type `A | B`
  - because object `{type: 'a', kind: 2, bOnly: number}` is not assignable to type `A` 
    - because `bOnly` is not a part of type `A`
  - because object `{type: 'a', kind: 2, bOnly: number}` is not assignable to type `B` 
    - because `2` cannot be assigned to `B.kind`

**Actual behavior:**
Ignores the error completely.

**Playground Link:** 
https://www.typescriptlang.org/play/#src=type%20A%20%3D%20%7B%0D%0A%20%20type%3A%20'a'%20%7C%20'b'%0D%0A%20%20kind%3A%201%20%7C%202%0D%0A%7D%0D%0A%0D%0Atype%20B%20%3D%20%7B%0D%0A%20%20type%3A%20'a'%0D%0A%20%20kind%3A%201%0D%0A%20%20bOnly%3A%20number%0D%0A%7D%0D%0A%0D%0Alet%20obj%3A%20A%20%7C%20B%0D%0A%0D%0Aobj%20%3D%20%7B%0D%0A%20%20type%3A%20'b'%2C%0D%0A%20%20kind%3A%202%2C%0D%0A%20%20bOnly%3A%2090%2C%20%2F%2F%20rightly%20complains%20that%20'bOnly'%20is%20not%20a%20part%20of%20type%20'A'%0D%0A%7D%0D%0A%0D%0Aobj%20%3D%20%7B%0D%0A%20%20type%3A%20'a'%2C%0D%0A%20%20kind%3A%202%2C%20%2F%2F%20does%20not%20complain%20that%20'2'%20cannot%20be%20assigned%20to%20'B.kind'%0D%0A%20%20bOnly%3A%2090%2C%20%2F%2F%20does%20not%20complain%20that%20'bOnly'%20is%20not%20a%20part%20of%20type%20'A'%0D%0A%7D

**Related Issues:** none

**Comment by weswigham** [comment]
> If you say variable can be of type A | B the compiler shouldn't let you assign something that has parts of type A and parts of type B.

You're describing an exclusive union - that is not what the `|` operator means in TS. Our `|` type is simple union - the sum of everything in both of its operands (including those types which satisfy both). The "exclusive union" behavior seen in, eg haskell, falls out of using the the same operator with nominal type semantics, however TS is a structural language, so "exclusive union" as a core operation makes less sense for us.
